### ISSUE TITLE
What is the variable `*standard-output*`?

### ISSUE THREAD

**Issue by mjambon** [BinaryAnalysisPlatform/bap#1057]
We just started getting an error involving an unknown variable `*standard-output*`. This variable is passed to our hook set on a `Primus.Interpreter.reading` observation. This behavior started very recently on the `master` branch, since pinning bap to commit 6efba4d (Feb 6) works. I see that the following code is not new and it initializes the environment:
```ocaml
    let setup_standard_channels =
      let set name descr =
        Lib.addr_width >>= fun width ->
        let v = Var.create name (Type.imm width) in
        Value.of_int ~width descr >>= Env.set v in
      Machine.sequence [
        set "*standard-input*" 0;
        set "*standard-output*" 1;
        set "*error-output*" 2; (* CL name *)
        set "*standard-error*" 2; (* conventional name *)
      ]
```
In our code, we check that we know about any variable that's being read. I'd like to understand why we're encountering a read to the `*standard-output*` variable and how the same effect was achieved (or not) before. Thanks in advance!

**Comment by ivg** [comment]
The `*standard-output*` variable is a Primus Lisp runtime variable that denotes the standard output channel. It is there for more than two years and nothing was changed in this code since that time (modulo #1049, that shouldn't directly affect this case).

This variable is used to model (to stub) IO operations, like read/write/gets/puts etc. You shall see this variable used when there is a write to standard output. If a variable is uninitialized, then it looks like that you're not loading the lisp plugin (and the code you've referenced is not running), but I doublt that could be true :) Besides, are you sure you're not getting a type error? What's the real error that happens? Or that's not Primus who is complaining and you just trying to understand why you didn't see this variable before? (For the latter the answer is that you didn't hit any function that writes to the output channel).

**Comment by mjambon** [comment]
I don't know much about Primus - I didn't write the code that uses it. Perhaps the first step is to produce a Primus trace as you suggest. How would I go about it? Is the best source of documentation about Primus the [mli file](https://github.com/BinaryAnalysisPlatform/bap/blob/master/lib/bap_primus/bap_primus.mli)?

**Comment by ivg** [comment]
We have the `primus-print` plugin that let you print and analyze primus observations (traces). So, in general, you can just do `--primus-print-observation=<list-of-events-that-are-interesting-to-you>`, e.g., for me the following list is usually a good start,

```
--primus-print-obs=pc-changed,read,written,loaded,stored,call,call-return,exception,enter-pos,lisp-message --primus-print-output=trace
```

Where `trace` is the name of file where the observations will be stored. You can also just say `all` to name all possible observations (and remove some observations from the list using `-<obs-name>`). This is all written in the bap manpage, `man bap`, so refer to it for details.

If you're not using bap as a command-line utility, then you can pass those options to `Bap_main.init` via the argv parameter or even use a configuration file.

There is also the [programmatical interface][1], which you can use to build traces and even integrate them with your logging facilities. In short, for each observation there is a provider (the list of all providers could be get via the `list_providers` function), then you can use `Provider.name p` which will be the name of the observation provided (e.g., `call`, `call-return`, `read`, `write` and so on) and the stream of actual observations (`Provider.stream`) where each element is encoded as Sexp.t and then print it in whatever place you like (and filter if you need), e.g. (careful, untested),
```ocaml
let print_primus_trace_to_formatter ppf = 
  Primus.Observation.list_providers () |>
  List.iter ~f:(fun p -> 
      Stream.observe (Primus.Observation.Provider.data p) @@ fun data ->
      Format.fprintf ppf "(%s %a) (Primus.Provider.name p) Sexp.pp data)
```


[1]: 
http://binaryanalysisplatform.github.io/bap/api/master/bap-primus/Bap_primus/Std/Primus/Observation/index.html

**Comment by ivg** [comment]
any updates on that? Were you able to obtain the trace?

**Comment by mjambon** [comment]
I haven't worked on it yet. I'll you posted.

**Comment by mjambon** [comment]
ok, so I'm back on it.

I must say that I'm not able to make sense of your suggestions. We're using bap as library, so I'm discarding the first suggestion to "pass" `--primus-print-obs=...` because it looks like a command-line argument and that's not how we're using bap. I looked at the programmatical interface that you mentioned (Primus.Observation) and the snippet you provided above, and I don't understand either. The following code that you posted doesn't compile as is, and I don't know where it would be called or what it does. I won't copy-paste code I don't understand, so I suggest providing a tracing functionality out of the box, as a preliminary to working on this bug. What would work is a simple function with obvious usage that would enable or disable tracing. It should come with reasonable defaults so that a user could just turn this on and get useful output for bug reports.

**Comment by ivg** [comment]
Ok, here is the corrected code with the full instructions:

```ocaml
open Core_kernel
open Bap_primus.Std
open Bap_future.Std

let print_primus_trace_to_formatter ppf =
  Primus.Observation.list_providers () |>
  List.iter ~f:(fun p ->
      Stream.observe (Primus.Observation.Provider.data p) @@ fun data ->
      Format.fprintf ppf "(%s %a)@\n"
        (Primus.Observation.Provider.name p) Sexp.pp data)

let () = print_primus_trace_to_formatter Format.std_formatter
```

you can test that it works by doing
```
bapbuild -pkg bap-primus trace.plugin
bapbundle install trace.plugin
```
assuming that you put the code above into the `trace.ml` file.


Also, even when using bap as a library, you can pass options to it using the `argv` parameter of `Bap_main.init`.

**Comment by mjambon** [comment]
Thank you! We're getting the following trace:
```
00:00:01.602 [12009] [seed 5d41402a] (init ())
00:00:01.602 [12009] [seed 5d41402a] (linker-exec syscall)
00:00:01.602 [12009] [seed 5d41402a] (interrupt 45)
00:00:01.602 [12009] [seed 5d41402a] (reading EBX)
00:00:01.602 [12009] [seed 5d41402a] (read (EBX 0:32u#27))
00:00:01.602 [12009] [seed 5d41402a] (reading EAX)
00:00:01.602 [12009] [seed 5d41402a] (read (EAX 0x804C000:32u#25))
00:00:01.602 [12009] [seed 5d41402a] (fini ())
00:00:01.603 [12009] [seed 5d41402a] (reading *standard-output*)
00:00:01.603 [12009] [seed 5d41402a] (exception "missing var: *standard-output*")
00:00:01.603 [12009] [seed 5d41402a] Error: (Missing_state_element"missing var: *standard-output*")
00:00:01.631 [12006] [seed 5d41402a] Error while dealing with child process 12009: Exception in child process: Uncaught exception in child process seed 5d41402a 12009:
00:00:01.631 [12006] [seed 5d41402a] Invalid_argument: output_value: abstract value (Custom)
```
The last two lines seem to indicate a problem with printing an exception in the parent process (note the PID 12006 vs. 12009). It looks unrelated.

Don't know it that helps. Let me see if I can produce a trace using the slightly earlier version of bap that doesn't exhibit the problem (commit 6efba4d).

**Comment by ivg** [comment]
The first trace is benign and is what should happen all the time. The `fini` observation triggers the [fini][1] method in `stdio` module, which flushes the channels. It is hard to tell why it didn't happen before, mostly because of the order of evaluation between different components competing for init/fini observations, see the Robustness section in #1075. After it is merged, it should happen always. 

To summarize, these are valid variables and they are read every time a computation finishes. 

[1]: https://github.com/BinaryAnalysisPlatform/bap/blob/ebb53a54d5b46669ff08ae133c215b5d2bf61c4a/plugins/primus_lisp/lisp/stdio.lisp#L113

**Comment by ivg** [comment]
closing this, as apparently there are no more questions.
